# `yasim` -- Yet Another SIMulator

Read simulator for Next- and Third-Generation Sequencing with Alternative Splicing and Differently Expressed Genes.

## WARNING

This simulator is still under active internal develop, so source code here **are for reference only**.

## Installation

You need Python intepreter (CPython implementation) >= 3.7 and latest [`setuptools`](https://setuptools.pypa.io/) to build this software. You are recommended to build the software in a virtual environment provided by [`virtualenv`](https://virtualenv.pypa.io), etc.

Build the simulator using:

```shell
python3 setup.py sdist
pip install dist/yasim-0.2.7-hotfix1.tar.gz
```

## Using `yasim`

List all available subcommands: `yasim lscmd`.

An example using hg38:

```shell
yasim generate_as_events -f hg38.fa -g hg38.gtf -o hg38_as.gtf
yasim generate_depth -g hg38_as.gtf -o hg38_depth.tsv -d 100
yasim transcribe -f hg38.fa -g hg38_as.gtf -o hg38_trans.fa
yasim pbsim2 \
    -F hg38_trans.fa.d \
    -o hg38_pbsim \
    -d hg38_depth.tsv \
    --hmm_model P4C2 \
    --exename PATH_TO_PBSIM2 \
    -j 40
```

## Usage

### Generate Alternative Splicing Events: `generate_as_events`

usage: `python3 -m yasim generate_as_events [-h] -f [FASTA] -g [GTF] -o [OUT]`

optional arguments:

- `-h`, `--help` show this help message and exit
- `-f [FASTA]`, `--fasta [FASTA]` Reference genome, in FASTA format
- `-g [GTF]`, `--gtf [GTF]` Reference genome, in GTF format
- `-o [OUT]`, `--out [OUT]` Output GTF

### Generate Sequencing Depth of Gene: `generate_depth`

usage: `python3 -m yasim generate_depth [-h] -f [FASTA] -g [GTF] -o [OUT]`

Arguments:

- `-h`, `--help` show this help message and exit
- `-g [GTF]`, `--gtf [GTF]` Reference genome, in GTF format
- `-o [OUT]`, `--out [OUT]` Output TSV
- `-d [MU]`, `--mu [MU]` Average depth

### Transcribe GTF to FASTA: `transcribe`

- `-h`, `--help` show this help message and exit
- `-f [FASTA]`, `--fasta [FASTA]` Reference genome, in FASTA format
- `-g [GTF]`, `--gtf [GTF]` GTF to be transcribed
- `-o [OUT]`, `--out [OUT]` Output FASTA

### Simulate: Use `badread` for Example

- `-h`, `--help` show this help message and exit
- `-F [FASTAS]`, `--fastas [FASTAS]` Directory of transcribed DGE FASTAs from `transcribe` step
- `-g [GTF]`, `--gtf [GTF]` GTF to be transcribed
- `-o [OUT]`, `--out [OUT]` Output transcript prefix
- `-d [DEPTH]`, `--depth [DEPTH]` Depth generated by `generate_depth` step
- `-m [MODEL_NAME]`, `--model_name [MODEL_NAME]`. Used model. Should be one of  -m `nanopore2018`,`nanopore2020`,`pacbio2016`,`verybad`,`verynice`
- `-e [EXENAME]`, `--exename [EXENAME]` Executable name or absolute path.
- `-j [JOBS]`, `--jobs [JOBS]` Number of threads used.

## Notice

To simulate using simulators like `pbsim`, `pbsim2`, etc., you are recommended to provide the **ABSOLUTE PATH** to the executable using `-e` argument.

## NEWS

- 2022-11-09: The `yasim.commonutils` and `yasim.bioutils` package would be separated into a new package (still under internal development) and `yasim` of later versions would depend on it.
